SG-21012 Add event MD files and replace shotgun instances (#182)
shotgunsoftware/developer.shotgunsoftware.com: docs/en/event-daemon/event-daemon-configuration.md

```

<a id="Plugin_Settings"></a>
### Plugin Settings

You will need to tell the {% include product %}EventDaemon where to look for plugins to run. Under the `[plugins]` section replace the default token with the correct  value for `paths`.

You can specify multiple locations (which may be useful if you have multiple departments or repositories using the daemon). The value here must be a full path to a readable existing directory.

Example

```
paths: /usr/local/shotgun/{% include product %}Events/plugins
```

When you're first getting started, a good plugin to test with is the `logArgs.py` plugin located in the `/usr/local/shotgun/{% include product %}Events/src/examplePlugins` directory. Copy that into the plugins folder you specified and we'll use that for testing things.

<a id="Location_of_shotgunEventDaemon_conf"></a>
### Location of shotgunEventDaemon.conf

By default, the daemon will look for the shotgunEventDaemon.conf file in the same directory that {% include product %}EventDaemon.py is in, and in the `/etc` directory. If you need to put the conf file in another directory, it's recommended you create a symlink to it from the current directory.

> {% include info title="Note" content="If for some reason the above won't work for you, the search paths for the config file are located in the `_getConfigPath()` function at the bottom of the `{% include product %}EventDaemon.py` script" %}

Reviewer: ```suggestion
{% include info title="Note" content="If for some reason the above won't work for you, the search paths for the config file are located in the `_getConfigPath()` function at the bottom of the `shotgunEventDaemon.py` script" %}
```
```suggestion
{% include info title="Note" content="If for some reason the above won't work for you, the search paths for the config file are located in the `_getConfigPath()` function at the bottom of the `{% include product %}EventDaemon.py` script" %}
```

pagename: event-daemon-configuration
lang: en
---

# Configuration

The following guide will help you configure {% include product %}Events for your studio.

Most of the configuration for {% include product %}Events is controlled by the `{% include product %}EventDaemon.conf`
file. In this file, you'll find several settings you can modify to match your
needs. Most of them have defaults that will work fine for most studios, however,
there are some settings that must be configured (specifically, your {% include product %}
server URL, script name, and application key so the {% include product %}EventDaemon can connect to your {% include product %} server).

> {% include info title="Note" content="**Windows note:** Windows users will need to change all the paths in the configuration file for Windows equivalents. We suggest keeping all paths, including logging, under one single location for the sake of simplicity. This documentation tends to refer to `C:\shotgun\shotgunEvents` when mentioning Windows paths." %}

Reviewer: ```suggestion
{% include info title="Note" content="**For Windows:** Windows users will need to change all the paths in the configuration file for Windows equivalents. We suggest keeping all paths, including logging, under one single location for the sake of simplicity. This documentation tends to refer to `C:\shotgun\shotgunEvents` when mentioning Windows paths." %}
```

```

<a id="Plugin_Settings"></a>
### Plugin Settings

You will need to tell the {% include product %}EventDaemon where to look for plugins to run.
Under the `[plugins]` section replace the default token with the correct  value
for `paths`.

You can specify multiple locations (which may be useful if you have multiple
departments or repositories using the daemon). The value here must be a full
path to a readable existing directory.

Example

```
paths: /usr/local/shotgun/shotgunEvents/plugins
```

When you're first getting started, a good plugin to test with is the
`logArgs.py` plugin located in the
`/usr/local/shotgun/shotgunEvents/src/examplePlugins` directory. Copy that into
the plugins folder you specified and we'll use that for testing things.

<a id="Location_of_shotgunEventDaemon_conf"></a>
### Location of {% include product %}EventDaemon.conf

By default, the daemon will look for the {% include product %}EventDaemon.conf file in the same directory that {% include product %}EventDaemon.py is in, and in the `/etc` directory.
If you need to put the conf file in another directory, it's recommended you 
create a symlink to it from the current directory.

> {% include info title="Note" content="**Note:** If for some reason the above won't work for you, the search paths for the
> config file are located in the `_getConfigPath()` function at the bottom of
> the `shotgunEventDaemon.py` script" %}

Reviewer: ```suggestion
{% include info title="Note" content="If for some reason the above won't work for you, the search paths for the config file are located in the `_getConfigPath()` function at the bottom of the `shotgunEventDaemon.py` script" %}
```

> {% include info title="Note" content="**Windows Note:** The `/etc` doesn't exist on Windows so the configuration file should be put in the same directory as the Python files." %}

Reviewer: ```suggestion
{% include info title="Note" content="**For Windows** The `/etc` doesn't exist on Windows so the configuration file should be put in the same directory as the Python files." %}
```

<a id="Testing_the_Daemon"></a>
## Testing the Daemon

Daemons can be difficult to test since they run in the background. There isn't
always an obvious way to see what they're doing. Lucky for us, the 
{% include product %}EventDaemon has an option to run it as a foreground process. Now that
we have done the minimum required setup, let's go ahead and test the daemon
and see how things go.

> {% include info title="Note" content="**Note:** The default values used here may require root access (for example, to write to the/var/log directory). The examples provided use are run using `sudo` to 
> accommodate this." %}

Reviewer: ```suggestion
{% include info title="Note" content="The default values used here may require root access (for example, to write to the/var/log directory). The examples provided use are run using `sudo` to accommodate this." %}
```

> {% include info title="Note" content="**Note:** The default values used here may require root access (for example, to write to the/var/log directory). The examples provided use are run using `sudo` to 
> accommodate this." %}

Reviewer: ```suggestion
{% include info title="Note" content="The default values used here may require root access (for example, to write to the/var/log directory). The examples provided use are run using `sudo` to accommodate this." %}
```

```
$ sudo ./shotgunEventDaemon.py foreground
INFO:engine:Using Shotgun version 3.0.8
INFO:engine:Loading plugin at /usr/local/shotgun/shotgunEvents/src/examplePlugins/logArgs.py
INFO:engine:Last event id (248429) from the Shotgun database.
```

You should see the lines above when you start the script (some of the details
may differ obviously). If you get any errors, the script will terminate since we opted to run it in the foreground we'll see that happen. Some common errors are displayed below if you get stuck.

The `logArgs.py` plugin simply takes the event that occurred in {% include product %} and passes it to the logger. Not very exciting but it's a simple way to ensure that
the script is running and the plugin is working. If you're at a busy studio,
you may have already noticed a rapid stream of messages flowing by. If not,
login to your {% include product %} server in your web browser and change some values or create something. You should see log statements printed out to your terminal window corresponding to the type of event you generated with your changes. 

> {% include info title="Note" content="NOTE: there are variables in the logArgs.py file that need to be filled in with appropriate values. '$DEMO_SCRIPT_NAMES$' and '$DEMO_API_KEY$' must be edited to contain the same values that were used in the shotgunEventDaemon.conf file in order for the logging to function correctly." %}

Reviewer: ```suggestion
{% include info title="Note" content="There are variables in the logArgs.py file that need to be filled in with appropriate values. '$DEMO_SCRIPT_NAMES$' and '$DEMO_API_KEY$' must be edited to contain the same values that were used in the shotgunEventDaemon.conf file in order for the logging to function correctly." %}
```

starting up the daemon and the daemon will process only new events created after startup. 

This file keeps track of the last event id for *each* plugin and stores this
information in pickle format.

```
eventIdFile: /var/log/shotgunEventDaemon.id
```

**logMode**

The logging mode can be set to one of two values:

- **0** = all log messages in the main log file
- **1** = one main file for the engine, one file per plugin

When using a value of **1**, the log messages generated by the engine itself
will be logged to the main log file specified by the `logFile` config setting.
Any messages logged by a plugin will be placed in a file named 
`plugin.<plugin_name>`.

```
logMode: 1
```

**logPath**

The path where to put log files (both for the main engine and plugin log files). The name of the main log file is controlled by the ``logFile`` 
setting below.

```
logPath: /var/log/shotgunEventDaemon
```

> {% include info title="Note" content="**Note:** The shotgunEventDaemon will have to have write permissions for this
> directory. In a typical setup, the daemon is set to run automatically when the
> machine starts up and is given root privileges at that point." %}

Reviewer: ```suggestion
{% include info title="Note" content="The shotgunEventDaemon will have to have write permissions for this directory. In a typical setup, the daemon is set to run automatically when the machine starts up and is given root privileges at that point." %}
```

- **error** If the callback failed or not. The value can be `False` or `True`.
- **delay** The duration in `DD:HH:MM:SS.micro_second` format of the delay between the event creation and the start of processing by the callback.

**conn_retry_sleep**

If the connection to {% include product %} fails, this is the number of seconds to wait until the connection is re-attempted. This allows for occasional network hiccups, 
server restarts, application maintenance, etc.

```
conn_retry_sleep = 60
```

**max_conn_retries**

Number of times to retry the connection before logging an error level message 
(which potentially sends an email if email notification is configured below). 

```
max_conn_retries = 5
```

**fetch_interval**

The number of seconds to wait before requesting new events after each batch of events is done being processed. This setting generally doesn't need to be adjusted.

```
fetch_interval = 5
```

<a id="Shotgun_Settings"></a>
### {% include product %} Settings

The following are settings related to your {% include product %} instance.

**server**

The URL for the {% include product %} server to connect to.

```
server: %(SG_ED_SITE_URL)s
```

> {% include info title="Note" content="**Note:** There is no default value here. Set the `SG_ED_SITE_URL` environment variable to 
> the URL for your Shotgun server (ie. https://awesome.shotgunstudio.com)" %}

Reviewer: ```suggestion
{% include info title="Note" content="There is no default value here. Set the `SG_ED_SITE_URL` environment variable to the URL for your Shotgun server (ie. https://awesome.shotgunstudio.com)" %}
```

**name**

The {% include product %} Script name the {% include product %}EventDaemon should connect with.

```
name: %(SG_ED_SCRIPT_NAME)s
```
        

> {% include info title="Note" content="**Note:** There is no default value here. Set the `SG_ED_SCRIPT_NAME` environment variable
> to the Script name for your Shotgun server (ie. `shotgunEventDaemon`)" %}

Reviewer: ```suggestion
{% include info title="Note" content="There is no default value here. Set the `SG_ED_SCRIPT_NAME` environment variable to the Script name for your Shotgun server (ie. `shotgunEventDaemon`)" %}
```

**key**

The {% include product %} Application Key for the Script name specified above.

```
key: %(SG_ED_API_KEY)s
```
        

> {% include info title="Note" content="**Note:** There is no default value here. Set the `SG_ED_API_KEY` environment variable to
> the Application Key for your Script name above (ie. 
> `0123456789abcdef0123456789abcdef01234567`)" %}

Reviewer: ```suggestion
{% include info title="Note" content="There is no default value here. Set the `SG_ED_API_KEY` environment variable to the Application Key for your Script name above (ie:`0123456789abcdef0123456789abcdef01234567`)" %}
```

**use_session_uuid**

Sets the session_uuid from every event in the {% include product %} instance to propagate in
any events generated by plugins. This will allow the {% include product %} UI to display
updates that occur as a result of a plugin.

```
use_session_uuid: True
```

- {% include product %} server v2.3+ is required for this feature.
- {% include product %} API v3.0.5+ is required for this feature.

> {% include info title="Note" content="**Note:** The Shotgun UI will *only* show updates live for the browser session that spawned the original event. Other browser windows with the same page open 
> will not see updates live." %}

Reviewer: ```suggestion
{% include info title="Note" content="The Shotgun UI will *only* show updates live for the browser session that spawned the original event. Other browser windows with the same page open will not see updates live." %}
```

<a id="Plugin_Settings_details"></a>
### Plugin Settings

**paths**

A comma-delimited list of complete paths where the framework should look for plugins to load. Do not use relative paths.

```
paths: /usr/local/shotgun/plugins
```

> {% include info title="Note" content="**Note:** There is no default value here. You must set the value to the location(s) of your plugin files (ie.
> `/usr/local/shotgun/shotgunEvents/plugins` or `C:\shotgun\shotgunEvents\plugins` on Windows)" %}

Reviewer: ```suggestion
{% include info title="Note" content="There is no default value here. You must set the value to the location(s) of your plugin files (ie:`/usr/local/shotgun/shotgunEvents/plugins` or `C:\shotgun\shotgunEvents\plugins` on Windows)" %}
```

<a id="Email_Settings"></a>
### Email Settings

These are used for error reporting because we figured you wouldn't constantly be
tailing the log and would rather have an active notification system.

Any error above level 40 (ERROR) will be reported via email if all of the
settings are provided below.

All of these values must be provided for there to be email alerts sent out.

**server**

The server that should be used for SMTP connections. The username and password
values can be uncommented to supply credentials for the SMTP connection. If 
your server does not use authentication, you should comment out the settings 
for `username` and `password`

```
server: smtp.yourdomain.com
```
        

> {% include info title="Note" content="**Note:** There is no default value here. You must replace the smtp.yourdomain.com
> token with the address of your SMTP server (ie. `smtp.mystudio.com`)." %}

Reviewer: ```suggestion
{% include info title="Note" content="There is no default value here. You must replace the smtp.yourdomain.com token with the address of your SMTP server (ie. `smtp.mystudio.com`)." %}
```

**username**

If your SMTP server requires authentication, uncomment this line and make sure you
have configured the `SG_ED_EMAIL_USERNAME` environment variable with the username
required to connect to your SMTP server.

```
username: %(SG_ED_EMAIL_USERNAME)s
```

**password**

If your SMTP server requires authentication, uncomment this line and make sure you
have configured the `SG_ED_EMAIL_PASSWORD` environment variable with the password
required to connect to your SMTP server.

```
password: %(SG_ED_EMAIL_PASSWORD)s
```

**from**

The from address that should be used in emails.

```
from: support@yourdomain.com
```

> {% include info title="Note" content="**Note:** There is no default value here. You must replace`support@yourdomain.com`
> with a valid value (ie. `noreply@mystudio.com`)." %}

Reviewer: ```suggestion
{% include info title="Note" content="There is no default value here. You must replace`support@yourdomain.com` with a valid value (ie. `noreply@mystudio.com`)." %}
```

**to**

A comma-delimited list of email addresses to whom these alerts should be sent.

```
to: you@yourdomain.com
```

> {% include info title="Note" content="**Note:** There is no default value here. You must replace `you@yourdomain.com`
> with a valid value (ie. `shotgun_admin@mystudio.com`)." %}

Reviewer: ```suggestion
{% include info title="Note" content="There is no default value here. You must replace `you@yourdomain.com` with a valid value (ie. `shotgun_admin@mystudio.com`)." %}
```
